Reduce docker base image size

XN137 opened the pull request:
- clean up unneeded files in binder postBuild
- only copy notebook files
- add script for local docker runs

only saves a couple hundred mb compared to the released images:
```
chris@xanadu:~/code/nessie-demos$ docker images
REPOSITORY                                  TAG                                        IMAGE ID       CREATED         SIZE
r2d-2e1646306922                            latest                                     f6d44866c11c   4 hours ago     5.77GB
r2d-2e1646306500                            latest                                     f6d44866c11c   4 hours ago     5.77GB
projectnessie/nessie-binder-demos           e12478cfb3447af4bc26e535a9daffe374233aa1   c17f6a22dd49   4 hours ago     5.77GB
(...)
ghcr.io/projectnessie/nessie-binder-demos   7fb962b09182c5f60ac2f80bbabb693290f3f741   0f701f74995f   2 months ago    6.49GB
```
but better than nothing.
also the demos are now easier to use because the `notebooks` folder in the image does not contain unneeded files anymore, just the notebooks.

Instead of just documenting the local docker image build process for testing purposes i wrote a script that automates it.

XN137 commented on docker/binder/postBuild:
```diff
@@ -33,3 +33,7 @@ python -c "import utils;utils.fetch_spark()"
+
+# clean up files that are not needed in the docker image
+rm -rf ~/.cache/pip/
+rm -rf /tmp/* || true
```
some folders fail to remove cleanly

omarsmak commented on build_run_local_docker.sh:
```diff
@@ -0,0 +1,48 @@
+# limitations under the License.
+#
+
+if ! python -m pip list | grep jupyter-repo2docker; then
```
How about something like `if ! command -v jupyter-repo2docker 2>/dev/null` ?

omarsmak reviewed (commented):
I think one of the biggest issue with the size that we had to copy Flink Jars into Pyflink, not sure what would be better way to do it 🤔

XN137 commented:
@omarsmak i've addressed all your comments

also found another way to reduce the image size, so we reduce it by ~1GB in total now:
```
REPOSITORY                                  TAG                                        IMAGE ID       CREATED             SIZE
r2d-2e1646392563                            latest                                     daa146619274   15 minutes ago      5.46GB
projectnessie/nessie-binder-demos           294ec5908c3b7aa970d99a2f5c7a669773cb591d   c95d9ca26b8a   16 minutes ago      5.46GB
(..)
projectnessie/nessie-binder-demos           e12478cfb3447af4bc26e535a9daffe374233aa1   14bf09fb62ef   2 hours ago         5.76GB
(..)
ghcr.io/projectnessie/nessie-binder-demos   7fb962b09182c5f60ac2f80bbabb693290f3f741   0f701f74995f   2 months ago        6.49GB
```

XN137 commented on .github/scripts/create_and_push_docker_image.sh:
```diff
@@ -15,50 +15,28 @@
-docker push "${INPUT_IMAGE_NAME}:latest"
-
-# Set the output docker tag we have from here
-echo "::set-output name=image_tag::${INPUT_IMAGE_TAG}"
```
was this used for anything?

omarsmak replied:
That was used before in the CI to pass the result to the next step. I think now we don't need and thus can be removed.

omarsmak commented on .github/scripts/create_and_push_docker_image.sh:
```diff
@@ -15,50 +15,28 @@
-  exit 1
-fi
+# ghcr.io/projectnessie/nessie-binder-demos:e12478cfb3447af4bc26e535a9daffe374233aa1
+DOCKER_IMAGE_NAME=$(echo "${INPUT_DOCKER_FULL_IMAGE_NAME}" | cut -f1 -d":")
```
Can we add a check here if the input docker image name has been passed?

XN137 replied:
ok i will add it back

omarsmak commented on .github/scripts/create_and_push_docker_image.sh:
```diff
@@ -15,50 +15,28 @@
-if [ -z "$INPUT_IMAGE_TAG" ]; then
-  echo "Input the image tag for the docker image as the second parameter"
-  exit 1
+if [ "$INPUT_PUSH" = 'push' ]; then
```
Do we need to check first if the second argument is set before the string comparison?

XN137 replied:
do we? right now the param is passed in from all callers.
and even if it is not this test would just be false (and so no pushing is done by default, which is fine?)

omarsmak replied:
Yeah all good, I was just not sure about it 👍

omarsmak commented on build_run_local_docker.sh:
```diff
@@ -0,0 +1,42 @@
+else
+  # build base image with the new tag locally
+  echo "base image tag has changed! - starting build..."
+  bash -ex .github/scripts/create_and_push_docker_image.sh "${DOCKER_FULL_IMAGE_NAME}" dontpush
```
Do we need here to pass dontpush argument? I think if is not passed means no push is needed.

XN137 replied:
since the file is still called `create_and_push_docker_image` i thought we should pass a param that makes it clear we are not pushing... i can also rename the file if you prefer that? then we dont need the param

omarsmak replied:
Yeah renaming the file and just not pass the push argument whenever the push is not needed would make more sense in my opinion.

XN137 replied:
ok done

omarsmak reviewed (changes requested):
Look good overall, just few comments. A question, does that linking working with pyflink? I presume all notebooks are working from the docker image?

XN137 commented:
> A question, does that linking working with pyflink? I presume all notebooks are working from the docker image?

yes to both answers, i was rerunning all the notebooks locally with all cells executing without errors

omarsmak reviewed (approved):
LGTM. Thanks @XN137!

XN137 commented:
workflow on main pushed the new image successfully:
```
Successfully pushed ghcr.io/projectnessie/nessie-binder-demos:294ec5908c3b7aa970d99a2f5c7a669773cb591d
```
https://github.com/orgs/projectnessie/packages/container/nessie-binder-demos/16188982?tag=294ec5908c3b7aa970d99a2f5c7a669773cb591d
